File: server/src/schemas/analysis.py
from __future__ import annotations

from pydantic import BaseModel, ConfigDict


class RecomendacaoResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    categoria: str
    descricao: str
    severidade: str


class MetricaResumida(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    tipo: str
    valor: float | None
    unidade: str | None


class AnalysisSimpleResponse(BaseModel):
    nota_geral: float | None = None
    feedback_ia: str | None = None
    recomendacoes: list[RecomendacaoResponse] = []
    metricas_resumidas: list[MetricaResumida] = []
    erro: str | None = None


class RawFrame(BaseModel):
    """Frame individual na resposta de dados brutos (US-018).

    ``keypoints`` é uma lista de 17 entradas, cada uma ``null`` ou
    ``[x, y, score]``. As métricas per-frame podem ser ``null`` quando os
    keypoints requeridos estão ausentes.
    """

    frame_idx: int
    keypoints: list[list[float] | None]
    angulo_joelho_esq: float | None = None
    angulo_joelho_dir: float | None = None
    angulo_cotovelo_esq: float | None = None
    angulo_cotovelo_dir: float | None = None
    inclinacao_tronco: float | None = None
    y_com: float | None = None


class SimetriaResponse(BaseModel):
    tcs: float | None = None
    joelho: float | None = None
    oscilacao: float | None = None


class TcsEvento(BaseModel):
    frame_idx: int
    tcs_ms: float


class OverstridingEvento(BaseModel):
    frame_idx: int
    overstriding_cm: float


class CadenciaEvento(BaseModel):
    frame_idx: int
    cadencia_spm: float


class EventosResponse(BaseModel):
    tcs_esq: list[TcsEvento] = []
    tcs_dir: list[TcsEvento] = []
    overstriding_esq: list[OverstridingEvento] = []
    overstriding_dir: list[OverstridingEvento] = []
    cadencia: list[CadenciaEvento] = []


class AnalysisRawResponse(BaseModel):
    fps: float | None = None
    frames: list[RawFrame] = []
    metricas_agregadas: list[MetricaResumida] = []
    simetria: SimetriaResponse = SimetriaResponse()
    eventos: EventosResponse = EventosResponse()
    erro: str | None = None

File: client/src/api/analysis.ts
import { api } from './client'

export interface RecomendacaoResponse {
  categoria: string
  descricao: string
  severidade: string
}

export interface MetricaResumida {
  tipo: string
  valor: number | null
  unidade: string | null
}

export interface AnalysisSimpleResponse {
  nota_geral: number | null
  feedback_ia: string | null
  recomendacoes: RecomendacaoResponse[]
  metricas_resumidas: MetricaResumida[]
  erro: string | null
}

export async function getSimpleAnalysisRequest(
  analysisId: number | string,
): Promise<AnalysisSimpleResponse> {
  const { data } = await api.get<AnalysisSimpleResponse>(
    `/api/analysis/${analysisId}/simple`,
  )
  return data
}

export interface RawFrame {
  frame_idx: number
  keypoints: (number[] | null)[]
  angulo_joelho_esq: number | null
  angulo_joelho_dir: number | null
  angulo_cotovelo_esq: number | null
  angulo_cotovelo_dir: number | null
  inclinacao_tronco: number | null
  y_com: number | null
}

export interface SimetriaRawResponse {
  tcs: number | null
  joelho: number | null
  oscilacao: number | null
}

export interface AnalysisRawResponse {
  fps: number | null
  frames: RawFrame[]
  metricas_agregadas: MetricaResumida[]
  simetria: SimetriaRawResponse
  erro: string | null
}

export async function getRawAnalysisRequest(
  analysisId: number | string,
): Promise<AnalysisRawResponse> {
  const { data } = await api.get<AnalysisRawResponse>(
    `/api/analysis/${analysisId}/raw`,
  )
  return data
}
